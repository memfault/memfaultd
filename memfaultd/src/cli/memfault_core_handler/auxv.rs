//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use scroll::Pread;
use std::fmt::Debug;

#[cfg(target_pointer_width = "64")]
pub type AuxvUint = u64;

#[cfg(target_pointer_width = "32")]
pub type AuxvUint = u32;

#[derive(Eq, PartialEq)]
pub struct Auxv<'a> {
    data: &'a [u8],
}

#[derive(Eq, PartialEq, Debug)]
pub struct AuxvEntry {
    pub key: AuxvUint,
    pub value: AuxvUint,
}

impl<'a> Auxv<'a> {
    pub fn new(data: &'a [u8]) -> Self {
        Self { data }
    }

    pub fn iter(&'a self) -> AuxvIterator<'a> {
        AuxvIterator::new(self)
    }

    pub fn find_value(&self, key: AuxvUint) -> Option<AuxvUint> {
        self.iter()
            .find(|a| a.key == key as AuxvUint)
            .map(|a| a.value)
    }
}

impl<'a> Debug for Auxv<'a> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.iter().collect::<Vec<_>>().fmt(f)
    }
}

pub struct AuxvIterator<'a> {
    auxv: &'a Auxv<'a>,
    offset: usize,
}

impl<'a> AuxvIterator<'a> {
    fn new(auxv: &'a Auxv<'a>) -> Self {
        Self { auxv, offset: 0 }
    }
}

impl<'a> Iterator for AuxvIterator<'a> {
    type Item = AuxvEntry;

    fn next(&mut self) -> Option<Self::Item> {
        let mut read = || self.auxv.data.gread::<AuxvUint>(&mut self.offset);
        match (read(), read()) {
            (Ok(key), Ok(value)) => Some(AuxvEntry { key, value }),
            _ => None,
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use rstest::rstest;
    use scroll::IOwrite;
    use std::io::Cursor;

    #[rstest]
    // Empty auxv:
    #[case(
        vec![],
        vec![]
    )]
    // Happy path:
    #[case(
        vec![1, 2, 3, 4],
        vec![AuxvEntry { key: 1, value: 2 }, AuxvEntry { key: 3, value: 4 }]
    )]
    // Partial entry at the end is ignored:
    #[case(
        vec![1, 2, 3],
        vec![AuxvEntry { key: 1, value: 2 }]
    )]
    fn test_auxv_iterator(#[case] values: Vec<AuxvUint>, #[case] expected: Vec<AuxvEntry>) {
        let buffer = make_fixture(values);
        let auxv = Auxv::new(buffer.as_slice());
        assert_eq!(auxv.iter().collect::<Vec<_>>(), expected);
    }

    #[test]
    fn test_auxv_find_value() {
        let buffer = make_fixture(vec![1, 2]);
        let auxv = Auxv::new(buffer.as_slice());
        assert!(auxv.find_value(1).is_some());
        assert!(auxv.find_value(9).is_none());
    }

    fn make_fixture(values: Vec<AuxvUint>) -> Vec<u8> {
        let mut cursor = Cursor::new(vec![]);
        for value in values {
            cursor.iowrite::<AuxvUint>(value).unwrap();
        }

        cursor.into_inner()
    }
}
