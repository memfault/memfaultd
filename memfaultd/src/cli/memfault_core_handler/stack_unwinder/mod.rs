//
// Copyright (c) Memfault, Inc.
// See License.txt for details
mod eh_frame_finder;
mod stacktrace_format;
mod unwind_handler;
mod unwinder;

pub use eh_frame_finder::EhFrameFinderImpl;
pub use unwind_handler::UnwindHandler;
pub use unwinder::{UnwindFrameContext, Unwinder};
