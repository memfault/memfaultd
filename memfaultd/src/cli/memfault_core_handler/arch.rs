//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::{collections::HashMap, mem::size_of};

use super::stack_unwinder::UnwindFrameContext;
use crate::cli::memfault_core_handler::elf;

use cfg_if::cfg_if;
use gimli::{Register, Vendor};

cfg_if! {
    if #[cfg(target_arch = "aarch64")] {
        pub use libc::user_regs_struct as ElfGRegSet;
        use gimli::AArch64;
        pub fn get_stack_pointer(regs: &ElfGRegSet) -> usize {
            regs.sp as usize
        }
        pub fn get_program_counter(regs: &ElfGRegSet) -> usize {
            regs.pc as usize
        }
        pub fn get_return_register(regs: &HashMap<Register, usize>) -> Option<usize> {
            // Register X30 is the LR on AArch64
            regs.get(&AArch64::X30).copied()
        }
        pub const fn return_register_idx() -> Register {
            AArch64::X30
        }
        pub fn set_stack_pointer(regs: &mut HashMap<Register, usize>, sp: usize)  {
            regs.insert(AArch64::SP, sp);
        }
        pub use elf::header::EM_AARCH64 as ELF_TARGET_MACHINE;
        pub use elf::header::ELFCLASS64 as ELF_TARGET_CLASS;
        impl From<&ElfGRegSet> for UnwindFrameContext {
            fn from(regs: &ElfGRegSet) -> Self {
                let mut ctx = UnwindFrameContext::default();
                ctx.registers.insert(AArch64::X0, regs.regs[0] as usize);
                ctx.registers.insert(AArch64::X1, regs.regs[1] as usize);
                ctx.registers.insert(AArch64::X2, regs.regs[2] as usize);
                ctx.registers.insert(AArch64::X3, regs.regs[3] as usize);
                ctx.registers.insert(AArch64::X4, regs.regs[4] as usize);
                ctx.registers.insert(AArch64::X5, regs.regs[5] as usize);
                ctx.registers.insert(AArch64::X6, regs.regs[6] as usize);
                ctx.registers.insert(AArch64::X7, regs.regs[7] as usize);
                ctx.registers.insert(AArch64::X8, regs.regs[8] as usize);
                ctx.registers.insert(AArch64::X9, regs.regs[9] as usize);
                ctx.registers.insert(AArch64::X10, regs.regs[10] as usize);
                ctx.registers.insert(AArch64::X11, regs.regs[11] as usize);
                ctx.registers.insert(AArch64::X12, regs.regs[12] as usize);
                ctx.registers.insert(AArch64::X13, regs.regs[13] as usize);
                ctx.registers.insert(AArch64::X14, regs.regs[14] as usize);
                ctx.registers.insert(AArch64::X15, regs.regs[15] as usize);
                ctx.registers.insert(AArch64::X16, regs.regs[16] as usize);
                ctx.registers.insert(AArch64::X17, regs.regs[17] as usize);
                ctx.registers.insert(AArch64::X18, regs.regs[18] as usize);
                ctx.registers.insert(AArch64::X19, regs.regs[19] as usize);
                ctx.registers.insert(AArch64::X20, regs.regs[20] as usize);
                ctx.registers.insert(AArch64::X21, regs.regs[21] as usize);
                ctx.registers.insert(AArch64::X22, regs.regs[22] as usize);
                ctx.registers.insert(AArch64::X23, regs.regs[23] as usize);
                ctx.registers.insert(AArch64::X24, regs.regs[24] as usize);
                ctx.registers.insert(AArch64::X25, regs.regs[25] as usize);
                ctx.registers.insert(AArch64::X26, regs.regs[26] as usize);
                ctx.registers.insert(AArch64::X27, regs.regs[27] as usize);
                ctx.registers.insert(AArch64::X28, regs.regs[28] as usize);
                ctx.registers.insert(AArch64::X29, regs.regs[29] as usize);
                ctx.registers.insert(AArch64::X30, regs.regs[30] as usize);
                ctx.registers.insert(AArch64::SP, regs.sp as usize);
                ctx.registers.insert(AArch64::PC, regs.pc as usize);

                ctx
            }
        }
    } else if #[cfg(target_arch = "x86_64")] {
        pub use libc::user_regs_struct as ElfGRegSet;
        use gimli::X86_64;
        pub fn get_stack_pointer(regs: &ElfGRegSet) -> usize {
            regs.rsp as usize
        }
        pub fn get_program_counter(regs: &ElfGRegSet) -> usize {
            regs.rip as usize
        }
        pub fn get_return_register(regs: &HashMap<Register, usize>) -> Option<usize> {
            regs.get(&X86_64::RA).copied()
        }
        pub const fn return_register_idx() -> Register {
            X86_64::RA
        }
        pub fn set_stack_pointer(regs: &mut HashMap<Register, usize>, sp: usize)  {
            regs.insert(X86_64::RSP, sp);
        }
        pub use elf::header::EM_X86_64 as ELF_TARGET_MACHINE;
        pub use elf::header::ELFCLASS64 as ELF_TARGET_CLASS;

        impl From<&ElfGRegSet> for UnwindFrameContext {
            fn from(regs: &ElfGRegSet) -> Self {
                let mut ctx = UnwindFrameContext::default();
                ctx.registers.insert(X86_64::RAX, regs.rax as usize);
                ctx.registers.insert(X86_64::RBX, regs.rbx as usize);
                ctx.registers.insert(X86_64::RCX, regs.rcx as usize);
                ctx.registers.insert(X86_64::RDX, regs.rdx as usize);
                ctx.registers.insert(X86_64::RSI, regs.rsi as usize);
                ctx.registers.insert(X86_64::RDI, regs.rdi as usize);
                ctx.registers.insert(X86_64::RBP, regs.rbp as usize);
                ctx.registers.insert(X86_64::RSP, regs.rsp as usize);
                ctx.registers.insert(X86_64::R8, regs.r8 as usize);
                ctx.registers.insert(X86_64::R9, regs.r9 as usize);
                ctx.registers.insert(X86_64::R10, regs.r10 as usize);
                ctx.registers.insert(X86_64::R11, regs.r11 as usize);
                ctx.registers.insert(X86_64::R12, regs.r12 as usize);
                ctx.registers.insert(X86_64::R13, regs.r13 as usize);
                ctx.registers.insert(X86_64::R14, regs.r14 as usize);
                ctx.registers.insert(X86_64::R15, regs.r15 as usize);
                ctx.registers.insert(X86_64::RFLAGS, regs.eflags as usize);
                ctx.registers.insert(X86_64::CS, regs.cs as usize);
                ctx.registers.insert(X86_64::FS, regs.fs as usize);
                ctx.registers.insert(X86_64::GS, regs.gs as usize);
                ctx.registers.insert(X86_64::SS, regs.ss as usize);
                ctx.registers.insert(X86_64::DS, regs.ds as usize);
                ctx.registers.insert(X86_64::ES, regs.es as usize);
                ctx.registers.insert(X86_64::FS_BASE, regs.fs_base as usize);
                ctx.registers.insert(X86_64::GS_BASE, regs.gs_base as usize);

                ctx
            }
        }
    } else if #[cfg(target_arch = "arm")] {
        pub use libc::user_regs as ElfGRegSet;
        pub fn get_stack_pointer(regs: &ElfGRegSet) -> usize {
            regs.arm_sp as usize
        }
        pub fn get_program_counter(regs: &ElfGRegSet) -> usize {
            regs.arm_pc as usize
        }
        pub fn set_stack_pointer(_regs: &mut HashMap<Register, usize>, sp: usize)  {
            todo!()
        }
        pub fn get_return_register(_regs: &HashMap<Register, usize>) -> Option<usize> {
            todo!()
        }
        pub const fn return_register_idx() -> Register {
            todo!()
        }
        pub use elf::header::EM_ARM as ELF_TARGET_MACHINE;
        pub use elf::header::ELFCLASS32 as ELF_TARGET_CLASS;

        impl From<&ElfGRegSet> for UnwindFrameContext {
            fn from(regs: &ElfGRegSet) -> Self {
                todo!()
            }
        }
    } else if #[cfg(target_arch = "x86")] {
        pub use libc::user_regs_struct as ElfGRegSet;
        pub fn get_stack_pointer(regs: &ElfGRegSet) -> usize {
            regs.esp as usize
        }
        pub fn get_program_counter(regs: &ElfGRegSet) -> usize {
            regs.eip as usize
        }
        pub fn set_stack_pointer(_regs: &mut HashMap<Register, usize>, sp: usize)  {
            todo!()
        }
        pub fn get_return_register(_regs: &HashMap<Register, usize>) -> Option<usize> {
            todo!()
        }
        pub const fn return_register_idx() -> Register {
            todo!()
        }
        pub use elf::header::EM_386 as ELF_TARGET_MACHINE;
        pub use elf::header::ELFCLASS32 as ELF_TARGET_CLASS;

        impl From<&ElfGRegSet> for UnwindFrameContext {
            fn from(regs: &ElfGRegSet) -> Self {
                todo!()
            }
        }
    }
    else {
        // Provide dummy symbols for unsupported architectures. This is preferable to
        // a compile time error, as we want to be able to compile memfaultd for all
        // architectures, but we don't need register access for all of them. Currently
        // these registers are only used to filter out stack memory from coredumps.
        pub struct ElfGRegSet;
        pub fn get_stack_pointer(_regs: &ElfGRegSet) -> usize {
            0
        }
        pub fn get_program_counter(_regs: &ElfGRegSet) -> usize {
            0
        }
        pub fn set_stack_pointer(_regs: &mut HashMap<Register, usize>, sp: usize)  {
            todo!()
        }
        pub fn get_return_register(_regs: &HashMap<Register, usize>) -> Option<usize> {
            None
        }
        pub const fn return_register_idx() -> Register {
            todo!()
        }
        impl From<&ElfGRegSet> for UnwindFrameContext {
            fn from(regs: &ElfGRegSet) -> Self {
                UnwindFrameContext::default()
            }
        }
    }
}

// Function definitions for coredump thread filter support. If the target architecture
// is not supported, these functions will always return false.
cfg_if! {
    if #[cfg(any(
        target_arch = "aarch64",
        target_arch = "arm",
        target_arch = "x86",
        target_arch = "x86_64"
    ))] {
        pub const fn coredump_thread_filter_supported() -> bool {
            true
        }
    } else {
        pub const fn coredump_thread_filter_supported() -> bool {
            false
        }
    }
}

cfg_if! {
    if #[cfg(target_arch = "aarch64")] {
        pub fn get_vendor() -> Vendor {
            Vendor::AArch64
        }
    } else {
        pub fn get_vendor() -> Vendor {
            Vendor::Default
        }
    }
}

// Function definitions for stacktrace support. If the target architecture is not
// supported, these functions will always return false.
cfg_if! {
    if #[cfg(any(
        target_arch = "aarch64",
        target_arch = "x86_64"
    ))] {
        pub const fn stacktrace_supported() -> bool {
            true
        }
    } else {
        pub const fn stacktrace_supported() -> bool {
            false
        }
    }
}

#[cfg(target_endian = "little")]
pub use elf::header::ELFDATA2LSB as ELF_TARGET_ENDIANNESS;

#[cfg(target_endian = "big")]
pub use elf::header::ELFDATA2MSB as ELF_TARGET_ENDIANNESS;

pub const WORD_SIZE: usize = size_of::<usize>();
