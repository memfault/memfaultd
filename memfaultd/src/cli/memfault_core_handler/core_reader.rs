//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::io::{Read, Seek, SeekFrom};

use crate::{cli::memfault_core_handler::elf, util::io::ForwardOnlySeeker};

use elf::header::Header;
use elf::program_header::{ProgramHeader, PT_NOTE};
use eyre::Result;

use super::elf_utils::{read_elf_header, read_program_headers};

pub trait CoreReader {
    /// Reads program headers from the input stream
    fn read_program_headers(&mut self) -> Result<Vec<ProgramHeader>>;

    /// Reads segment data from the input stream
    fn read_segment_data(&mut self, program_header: &ProgramHeader) -> Result<Vec<u8>>;

    /// Reads all note segments from the input stream
    fn read_all_note_segments<'a>(
        &mut self,
        program_headers: &'a [ProgramHeader],
    ) -> Vec<(&'a ProgramHeader, Vec<u8>)>;
}

// Reads ELF headers and segments from a core file
pub struct CoreReaderImpl<R: Read> {
    input_stream: ForwardOnlySeeker<R>,
    elf_header: Header,
}

impl<R: Read> CoreReader for CoreReaderImpl<R> {
    fn read_program_headers(&mut self) -> Result<Vec<ProgramHeader>> {
        self.input_stream
            .seek(SeekFrom::Start(self.elf_header.e_phoff as _))?;

        let mut program_headers =
            read_program_headers(&mut self.input_stream, self.elf_header.e_phnum as usize)?;

        // Sort, just in case the program headers are not sorted by offset.
        // Otherwise the read_segment_data() calls later may fail.
        program_headers.sort_by_key(|ph| ph.p_offset);

        Ok(program_headers)
    }

    fn read_segment_data(&mut self, program_header: &ProgramHeader) -> Result<Vec<u8>> {
        self.input_stream
            .seek(SeekFrom::Start(program_header.p_offset as _))?;

        let mut buf = vec![0; program_header.p_filesz as usize];
        self.input_stream.read_exact(&mut buf)?;

        Ok(buf)
    }

    fn read_all_note_segments<'a>(
        &mut self,
        program_headers: &'a [ProgramHeader],
    ) -> Vec<(&'a ProgramHeader, Vec<u8>)> {
        program_headers
            .iter()
            .filter_map(|ph| match ph.p_type {
                PT_NOTE => match self.read_segment_data(ph) {
                    Ok(data) => Some((ph, data)),
                    _ => None,
                },
                _ => None,
            })
            .collect::<Vec<_>>()
    }
}

impl<R: Read> CoreReaderImpl<R> {
    /// Creates an instance of `CoreReader` from an input stream
    pub fn new(input_stream: R) -> Result<Self> {
        let mut input_stream = ForwardOnlySeeker::new(input_stream);
        let elf_header = read_elf_header(&mut input_stream)?;

        Ok(CoreReaderImpl {
            input_stream,
            elf_header,
        })
    }

    pub fn elf_header(&self) -> Header {
        self.elf_header
    }
}

#[cfg(test)]
mod test {
    use std::io::Cursor;

    use super::*;

    use crate::cli::memfault_core_handler::arch::{
        ELF_TARGET_CLASS, ELF_TARGET_ENDIANNESS, ELF_TARGET_MACHINE,
    };
    use crate::cli::memfault_core_handler::test_utils::build_test_header;

    use rstest::rstest;
    use scroll::Pwrite;

    #[rstest]
    #[case(0)]
    #[case(1024)] // Test with padding between header and program headers
    fn test_read_program_headers(#[case] ph_offset: usize) {
        use elf::{header::SIZEOF_EHDR, program_header::SIZEOF_PHDR};

        let mut elf_header =
            build_test_header(ELF_TARGET_CLASS, ELF_TARGET_ENDIANNESS, ELF_TARGET_MACHINE);
        elf_header.e_phnum = 2;
        elf_header.e_phoff = (SIZEOF_EHDR + ph_offset).try_into().unwrap();
        let load_program_header = ProgramHeader {
            p_type: elf::program_header::PT_LOAD,
            p_vaddr: 0,
            ..Default::default()
        };
        let note_program_header = ProgramHeader {
            p_type: elf::program_header::PT_NOTE,
            p_offset: 0x1000,
            ..Default::default()
        };

        // Build ELF input stream
        let mut input_stream = vec![0; SIZEOF_EHDR + 2 * SIZEOF_PHDR + ph_offset];
        input_stream.pwrite(elf_header, 0).unwrap();
        input_stream
            .pwrite(load_program_header, SIZEOF_EHDR + ph_offset)
            .unwrap();
        input_stream
            .pwrite(note_program_header, SIZEOF_EHDR + SIZEOF_PHDR + ph_offset)
            .unwrap();

        // Verify headers are read correctly
        let mut reader = CoreReaderImpl::new(Cursor::new(input_stream)).unwrap();
        let program_headers = reader.read_program_headers().unwrap();
        assert_eq!(program_headers.len(), 2);
        assert_eq!(program_headers[0], load_program_header);
        assert_eq!(program_headers[1], note_program_header);
    }

    #[rstest]
    #[case(0, 1024)]
    #[case(1024, 1024)]
    fn test_read_segment_data(#[case] offset: usize, #[case] size: usize) {
        use elf::header::SIZEOF_EHDR;

        const TEST_BYTE: u8 = 0x42;

        let elf_header =
            build_test_header(ELF_TARGET_CLASS, ELF_TARGET_ENDIANNESS, ELF_TARGET_MACHINE);
        let offset = offset + SIZEOF_EHDR;
        let note_program_header = ProgramHeader {
            p_type: elf::program_header::PT_NOTE,
            p_offset: offset.try_into().unwrap(),
            p_filesz: size.try_into().unwrap(),
            ..Default::default()
        };

        let mut input_stream = vec![0u8; offset + size];
        input_stream.pwrite(elf_header, 0).unwrap();
        input_stream[offset..(offset + size)].fill(TEST_BYTE);

        let mut reader = CoreReaderImpl::new(Cursor::new(&input_stream)).unwrap();
        let segment_data = reader.read_segment_data(&note_program_header).unwrap();

        assert_eq!(segment_data, input_stream[offset..(offset + size)]);
    }
}
