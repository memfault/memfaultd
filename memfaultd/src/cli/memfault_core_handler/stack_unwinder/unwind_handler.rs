//
// Copyright (c) Memfault, Inc.
// See License.txt for details
//! This module provides a high level interface for unwinding stack traces from a core file.

use std::{
    io::{Read, Seek, Write},
    sync::mpsc::Receiver,
};

use crate::cli::memfault_core_handler::{
    arch::get_program_counter,
    core_elf_note::{ElfNote, ProcessStatusNote},
};

use eyre::{eyre, Result};
use log::debug;
use nix::sys::signal::Signal;
use serde_json::to_writer;

use super::{
    eh_frame_finder::EhFrameFinder,
    stacktrace_format::{StacktraceFormat, ThreadDescriptor},
    UnwindFrameContext, Unwinder,
};

/// The UnwindHandler is responsible holding everything needed to unwind a stack.
///
/// Its main purpose is to group everything need to unwind a stack trace in one place,
/// and provide a simple interface to do so. Additionally this makes it easier to test
/// the unwinding logic, as all of the inputs and outputs are generic.
pub struct UnwindHandler<P, O>
where
    P: Read + Seek,
    O: Write,
{
    proc_mem_stream: P,
    output_stream: O,
}

impl<P, O> UnwindHandler<P, O>
where
    P: Read + Seek,
    O: Write,
{
    pub fn new(proc_mem_stream: P, output_stream: O) -> Self {
        Self {
            proc_mem_stream,
            output_stream,
        }
    }

    pub fn build_stacktrace<E: EhFrameFinder>(
        &mut self,
        mut eh_frame_finder: E,
        elf_notes: &[ElfNote],
        cmd_line: String,
        core_handler_logs: Receiver<String>,
        crashing_tid: i32,
        signal: i32,
    ) -> Result<()> {
        let mut unwinder = Unwinder::new(&mut eh_frame_finder);

        // Collect all prstatus notes from the core file. These will act as the starting points
        // for unwinding each stack.
        let prstatus_notes = elf_notes
            .iter()
            .filter_map(|note| match note {
                ElfNote::ProcessStatus(s) => Some(s),
                _ => None,
            })
            .collect::<Vec<_>>();
        let signal = Signal::try_from(signal).map_err(|e| eyre!(e))?.to_string();

        // Get the stacktraces for each thread and write them to the output file.
        let threads = prstatus_notes
            .iter()
            .map(|prstatus_note| {
                (
                    prstatus_note.pr_pid,
                    self.unwind_stack(prstatus_note, &mut unwinder),
                )
            })
            .collect::<Vec<_>>();

        let thread_descriptors = threads
            .into_iter()
            .map(|(tid, pcs)| {
                let active = tid as i32 == crashing_tid;
                println!("Thread {} active: {}", tid, active);
                ThreadDescriptor::new(active, pcs)
            })
            .collect::<Vec<_>>();
        let symbol_descriptors = eh_frame_finder.get_symbol_file_descriptors()?;
        let trace_logs = core_handler_logs.try_iter().collect();
        let backtrace = StacktraceFormat::new(
            signal,
            cmd_line,
            symbol_descriptors,
            thread_descriptors,
            trace_logs,
        );

        to_writer(&mut self.output_stream, &backtrace)?;

        Ok(())
    }

    fn unwind_stack<E: EhFrameFinder>(
        &mut self,
        prstatus_note: &ProcessStatusNote,
        unwinder: &mut Unwinder<E>,
    ) -> Vec<usize> {
        let regs = &prstatus_note.pr_reg;
        let pc = get_program_counter(regs);

        let mut ctx = UnwindFrameContext::from(regs);
        if let Err(e) = unwinder.unwind_stack(pc, &mut ctx, &mut self.proc_mem_stream) {
            debug!("Stack unwinding aborted: {}", e);
        }

        ctx.pc_stack
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use std::{fs::File, path::PathBuf, sync::mpsc::channel};

    use crate::cli::memfault_core_handler::{
        core_elf_note::iterate_elf_notes,
        core_reader::{CoreReader, CoreReaderImpl},
        elf_utils::{read_elf_header, SectionMap},
        stack_unwinder::{eh_frame_finder::EhFrameInfo, stacktrace_format::SymbolFileDescriptor},
        test_utils::FakeProcMem,
    };

    use eyre::eyre;
    use insta::assert_json_snapshot;
    use serde_json::from_reader;

    // The crashing thread id from the simple exe core file.
    const CRASHING_TID: i32 = 3518029;

    #[test]
    fn test_unwind_stack() {
        let core_path = core_path();
        let core_file = File::open(core_path.clone()).unwrap();
        let mut core_reader = CoreReaderImpl::new(core_file).unwrap();
        let proc_mem = FakeProcMem::new_from_path(core_path).unwrap();

        let program_headers = core_reader.read_program_headers().unwrap();
        let all_notes = core_reader.read_all_note_segments(&program_headers);
        let parsed_notes = all_notes
            .iter()
            .flat_map(|(_, data)| iterate_elf_notes(data))
            .collect::<Vec<_>>();

        let eh_frame_finder = SimpleEhFrameFinder::new();

        let mut output_buf: Vec<u8> = Vec::new();
        let mut unwind_handler = UnwindHandler::new(proc_mem, &mut output_buf);
        let (tx, rx) = channel();
        tx.send("Ranger".to_string()).unwrap();
        tx.send("Daisy".to_string()).unwrap();
        tx.send("Carlton".to_string()).unwrap();
        unwind_handler
            .build_stacktrace(
                eh_frame_finder,
                &parsed_notes,
                "simple_exe".to_string(),
                rx,
                CRASHING_TID,
                11,
            )
            .unwrap();

        let stacktrace: StacktraceFormat = from_reader(&output_buf[..]).unwrap();
        assert_json_snapshot!(stacktrace);
    }

    struct SimpleEHFrameInfo {
        eh_frame_info: EhFrameInfo,
        start_addr: usize,
        end_addr: usize,
        path: String,
    }

    impl SimpleEHFrameInfo {
        fn new(
            path: PathBuf,
            start_addr: usize,
            end_addr: usize,
            compiled_base_addr: usize,
        ) -> Self {
            let mut file = File::open(path).unwrap();
            let header = read_elf_header(&mut file).unwrap();
            let mut section_map = SectionMap::new(file, &header).unwrap();

            let eh_frame = section_map
                .get_section(".eh_frame")
                .expect("Failed to find .eh_frame section")
                .unwrap();
            let eh_frame_hdr = section_map
                .get_section(".eh_frame_hdr")
                .expect("Failed to find .eh_frame_hdr section")
                .unwrap();

            let eh_frame_info =
                EhFrameInfo::new(eh_frame_hdr, eh_frame, compiled_base_addr, start_addr);

            SimpleEHFrameInfo {
                eh_frame_info,
                start_addr,
                end_addr,
                path: "test_path".to_string(),
            }
        }
    }

    impl From<&SimpleEHFrameInfo> for SymbolFileDescriptor {
        fn from(info: &SimpleEHFrameInfo) -> Self {
            SymbolFileDescriptor::new(
                info.start_addr,
                info.end_addr,
                "abcd1234".to_string(),
                info.eh_frame_info.compiled_base_addr,
                info.start_addr,
                info.path.clone(),
            )
        }
    }

    /// Represents a bespoke implementation of the `EhFrameFinder` trait for testing purposes.
    ///
    /// This allows us to load exactly the `.eh_frame` and `.eh_frame_hdr` sections we want to test
    /// the unwinder. currently this is just the ranges for the main executable and libc. All
    /// addresses were manually acquired through inspection of relevant ELF files, and the PC
    /// values in the coredump.
    struct SimpleEhFrameFinder {
        frame_info: Vec<SimpleEHFrameInfo>,
    }

    impl EhFrameFinder for SimpleEhFrameFinder {
        fn find_eh_frame(&mut self, pc: usize) -> Result<EhFrameInfo> {
            let eh_frame = self
                .frame_info
                .iter()
                .find(|info| pc >= info.start_addr && pc < info.end_addr)
                .map(|info| &info.eh_frame_info)
                .ok_or_else(|| eyre!("Failed to find eh_frame for pc: {:#x}", pc))?;

            Ok(eh_frame.clone())
        }

        fn get_symbol_file_descriptors(&self) -> Result<Vec<SymbolFileDescriptor>> {
            Ok(self.frame_info.iter().map(|info| info.into()).collect())
        }
    }

    impl SimpleEhFrameFinder {
        fn new() -> Self {
            let libc_simple_frame = SimpleEHFrameInfo::new(
                libc_ehframe_path(),
                0x75959a428000,
                0x75959a5bd000,
                0x28000,
            );
            let simple_exe_simple_frame =
                SimpleEHFrameInfo::new(simple_exe_path(), 0x63010b1cf000, 0x63010b1d0000, 0x1000);
            let frame_info = vec![libc_simple_frame, simple_exe_simple_frame];

            SimpleEhFrameFinder { frame_info }
        }
    }

    fn core_path() -> PathBuf {
        PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .join("src/cli/memfault_core_handler/fixtures/simple_executable/simple_exe_core.elf")
    }

    fn libc_ehframe_path() -> PathBuf {
        PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .join("src/cli/memfault_core_handler/fixtures/simple_executable/libc_ehframe.elf")
    }

    fn simple_exe_path() -> PathBuf {
        PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .join("src/cli/memfault_core_handler/fixtures/simple_executable/simple_exe.elf")
    }
}
