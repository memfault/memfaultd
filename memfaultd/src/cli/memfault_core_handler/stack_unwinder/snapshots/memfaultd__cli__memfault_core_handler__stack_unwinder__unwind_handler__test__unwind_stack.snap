---
source: memfaultd/src/cli/memfault_core_handler/stack_unwinder/unwind_handler.rs
expression: stacktrace
snapshot_kind: text
---
{
  "version": "1",
  "signal": "SIGSEGV",
  "cmdline": "simple_exe",
  "symbols": [
    {
      "pc_range": {
        "start": "0x75959a428000",
        "end": "0x75959a5bd000"
      },
      "build_id": "abcd1234",
      "compiled_offset": "0x28000",
      "runtime_offset": "0x75959a428000",
      "path": "test_path"
    },
    {
      "pc_range": {
        "start": "0x63010b1cf000",
        "end": "0x63010b1d0000"
      },
      "build_id": "abcd1234",
      "compiled_offset": "0x1000",
      "runtime_offset": "0x63010b1cf000",
      "path": "test_path"
    }
  ],
  "threads": [
    {
      "active": true,
      "pcs": [
        "0x63010b1cf161",
        "0x63010b1cf191",
        "0x63010b1cf1bf",
        "0x63010b1cf1f4",
        "0x75959a429d8f",
        "0x75959a429e3f",
        "0x63010b1cf084"
      ]
    }
  ],
  "core_handler_logs": [
    "Ranger",
    "Daisy",
    "Carlton"
  ]
}
